# ¡Hola Denisee! <a class="tocSkip"></a>

Mi nombre es David Bautista y tengo el gusto de revisar tu proyecto.

Si veo un error en la primera revisión solamente lo señalaré y dejaré que tú encuentres de qué se trata y cómo arreglarlo. Debo prepararte para que te desempeñes como especialista en Data, en un trabajo real, el responsable a cargo tuyo hará lo mismo. Si aún tienes dificultades para resolver esta tarea, te daré indicaciones más precisas en una siguiente iteración.

Te dejaré mis comentarios más abajo - **por favor, no los muevas, modifiques o borres**

Comenzaré mis comentarios con un resumen de los puntos que están bien, aquellos que debes corregir y aquellos que puedes mejorar. Luego deberás revisar todo el notebook para leer mis comentarios, los cuales estarán en rectángulos de color verde, amarillo o rojo como siguen:

<div class='alert alert-block alert-success'>
<b>Comentario de Reviewer</b> <a class='tocSkip'></a>

Muy bien! Toda la respuesta fue lograda satisfactoriamente.
</div>

<div class='alert alert-block alert-warning'>
<b>Comentario de Reviewer</b> <a class='tocSkip'></a>

Existen detalles a mejorar. Existen recomendaciones.
</div>

<div class='alert alert-block alert-danger'>

<b>Comentario de Reviewer</b> <a class='tocSkip'></a>

Se necesitan correcciones en el bloque. El trabajo no puede ser aceptado con comentarios en rojo sin solucionar.
</div>

Cualquier comentario que quieras agregar entre iteraciones de revisión lo puedes hacer de la siguiente manera:

<div class='alert alert-block alert-info'>
<b>Respuesta estudiante.</b> <a class='tocSkip'></a>
</div>

Mucho éxito en el proyecto!

<div class="alert alert-block alert-success">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

# Comentario General

Hola, Denisee. Te felicito por el desarrollo del proyecto. Lograste completar las diferentes secciones de manera correcta.

</div>


# Déjame escuchar música

Los datos están almacenados en el archivo `/datasets/music_project_en.csv`.

¡Ahora sí, manos al código!


## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Etapa 1.1. Necesitarás `pandas`, así que impórtalo.

In [1]:
# Importa pandas
import pandas as pd



<div class="alert alert-block alert-success">
<b>Comentario de Revisor </b> <a class="tocSkip"></a>

Perfecto, buen trabajo importando ``pandas``.  </div>


<details>
<summary>Haz clic para ver la pista</summary>

Para usar `pandas`, necesitas la palabra clave `import` seguida del nombre de la biblioteca y un alias.


Etapa 1.2. Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [2]:
# Lee el archivo y almacénalo en df
df = pd.read_csv('/datasets/music_project_en.csv')


<div class="alert alert-block alert-success">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

¡Perfecto! Muy buen trabajo cargando el dataset principal, un paso fundamental para el desarrollo del proyecto.
</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa la palabra clave `read_csv` para leer el archivo CSV. Asegúrate de incluir la ruta `/datasets/music_project_en.csv` y guarda el resultado en la variable `df`.

Etapa 1.3. Muestra las 10 primeras filas de la tabla:

In [3]:
# Obtén las 10 primeras filas de la tabla df
print(df.head(10))


     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

<div class="alert alert-block alert-success">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

Buen trabajo usando el método <code>head()</code>. Como tip, recuerda que en los notebooks de Jupyter se puede usar <code>display()</code> en lugar de <code>print()</code> para mostrar la salida de una tabla; esto genera una visualización mucho más estética. 
</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `head` para ver las primeras filas del DataFrame.

Etapa 1.4. Obtén la información general sobre la tabla con el método info().

In [4]:
# Obtén la información general sobre nuestros datos
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


<div class="alert alert-block alert-success">
<b>Comentario de Revisor  </b> <a class="tocSkip"></a>

Perfecto, buen trabajo usando el método ``info()``.   </div>




<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `info` para obtener detalles sobre columnas, tipos de datos y valores nulos.


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas que almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la canción;
- `'City'`: ciudad del usuario;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver dos problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. `Identifica tú mismo el segundo problema y escríbelo aquí.`
Respuesta: Un nombre de columna tiene espacio al incio.




### Escribe algunas observaciones por tu parte. Contesta a las siguientes preguntas: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos hay en las filas? ¿Cómo podemos saber qué almacenan las columnas?`

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`

Escribe aquí tus respuestas:

1. Todos los datos son tipo object. Podemos saber que datos almacenan las columnas con el método info de Pandas, y con la documentación proporcionada.

2. Considero que hace falta la fecha en la que el usuario escucha, ya que solo contamos con el día, sin la fecha no podremos identificar variaciones en los habitos de escucha de los usuarios. Tambien para lograr un resultado certero, es importante contar con la información de los 7 días de la semana.

3. Si, valores ausentes, uso de mayúsculas y minúsculas, datos con espacios al inicio.

<div class="alert alert-block alert-success">
<b>Comentario de Revisor  </b> <a class="tocSkip"></a>

Perfecto, buen trabajo.   </div>




## Etapa 2. Preprocesamiento de los datos <a id='data_preprocessing'></a>

Tu objetivo aquí es preparar los datos para analizarlos.
El primer paso es resolver los problemas con los encabezados. Después podemos avanzar a los valores ausentes y duplicados. ¡Empecemos!

Vamos a corregir el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Etapa 2.1. Muestra los encabezados de la tabla (los nombres de las columnas):

In [7]:
# Muestra los nombres de las columnas
df.columns

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')

<div class="alert alert-block alert-success">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

Perfecto, buen trabajo mostrando los encabezados del dataset.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa el atributo `.columns` para ver todos los nombres de las columnas en el DataFrame.


Vamos cambiar los encabezados de la tabla siguiendo las reglas estilísticas convencionales:
*   Todos los caracteres deben ser minúsculas.
*   Elimina los espacios.
*   Si el nombre tiene varias palabras, utiliza snake_case, es decir, añade un guion bajo ( _ ) entre las palabras en lugar de un espacio.



Etapa 2.2. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [5]:

# Bucle que itera sobre los encabezados y los pone todos en minúsculas
clean_list= []

for column in df.columns:
    clean_list.append(column.lower())
df.columns = clean_list
print(df.columns)
    

Index(['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day'], dtype='object')


<div class="alert alert-block alert-success">
<b>Comentario de Revisor </b> <a class="tocSkip"></a>

Excelente, buen trabajo usando el bucle con los métodos <code>append()</code> y <code>lower()</code> para modificar los encabezados.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa un `for` para recorrer `df.columns` y aplica `.lower()` a cada nombre. Luego, asigna la lista resultante de nuevo a `df.columns`.


Etapa 2.3. Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas y muestra los nombres de las columnas de nuevo:

In [9]:
# Bucle que itera sobre los encabezados y elimina los espacios
clean_list= []

for column in df.columns:
    clean_list.append(column.strip())
df.columns = clean_list
print(df.columns)


Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<div class="alert alert-block alert-success">
<b>Comentario de Revisor </b> <a class="tocSkip"></a>

Perfecto, buen trabajo usando el bucle con los métodos <code>append()</code> y <code>strip()</code> para modificar los encabezados.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa `.strip()` dentro del bucle para limpiar los nombres de columnas. Asegúrate de reasignar los nuevos nombres a `df.columns`.


Etapa 2.4. Necesitamos aplicar la regla de snake_case en la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [10]:
# Cambia el nombre de la columna "userid"
df.rename(columns={'userid':'user_id'}, inplace=True)
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<div class="alert alert-block alert-success">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

Excelente, buen trabajo aplicando el formato 🐍<strong>snake_case</strong> 🐍 a la variable <code>user_id</code>.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa el método `.rename()` con el parámetro `columns` para cambiar `userid` a `user_id`. No olvides establecer `inplace=True` o reasignar a `df`.

Etapa 2.5. Comprueba el resultado. Muestra los encabezados una vez más:

In [11]:
# Comprueba el resultado: lista de encabezados
df.columns

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

<details>
<summary>Haz clic para ver la pista</summary>

Usa `.columns` para volver a ver los nombres actualizados de las columnas.

### Valores ausentes <a id='missing_values'></a>
 Etapa 2.5. Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos para obtener el número de valores ausentes.

In [12]:
# Calcula el número de valores ausentes
df.isna().sum()

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64

<div class="alert alert-block alert-success">
<b>Comentario de Revisor </b> <a class="tocSkip"></a>

Perfecto, buen trabajo identificando la cantidad de valores nulos presentes en cada columna del dataset.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa `.isna()` para detectar los valores ausentes y combínalo con el método `.sum()` para contar cuántos hay por columna.


Etapa 2.6. Sustituye los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`.

1. Crea una lista llamada columns_to_replace que contenga los nombres de las columnas 'track', 'artist' y 'genre'.

2. Usa un bucle for para iterar sobre cada columna en columns_to_replace.

3. Dentro del bucle, sustituye los valores ausentes en cada columna con el string `'unknown'`.

In [13]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
columns_to_replace = ['track', 'artist', 'genre']

for column in columns_to_replace:
    df[column] = df[column].fillna('unknown')


<div class="alert alert-block alert-success">
<b>Comentario de Revisor </b> <a class="tocSkip"></a>

Excelente, buen trabajo usando el bucle con el método <code>fillna()</code> para solucionar el problema de los valores nulos dentro del dataset.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Crea una lista con los nombres de las columnas y usa un `for` para recorrerla. Dentro del bucle, aplica `.fillna('unknown')` a cada una.

Etapa 2.7. Ahora comprueba el resultado para asegurarte de que no falten valores ausentes por reemplazar en el conjunto de datos. Para ello, cuenta los valores ausentes una vez más.

In [14]:
# Cuenta los valores ausentes
df.isna().sum()

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64

<div class="alert alert-block alert-success">
<b>Comentario de Revisor </b> <a class="tocSkip"></a>

Perfecto, buen trabajo. 
</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa `isna()` con `sum()` otra vez para confirmar que ya no quedan valores nulos en las columnas modificadas.

### Duplicados <a id='duplicates'></a>
Etapa 2.8. Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos para obtener la cantidad de duplicados explícitos.

In [15]:
# Cuenta los duplicados explícitos
df.duplicated().sum()

3826

<div class="alert alert-block alert-success">
<b>Comentario de Revisor </b> <a class="tocSkip"></a>

Perfecto, buen trabajo identificando la cantidad de registros duplicados explicitos dentro del dataset.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa un método que detecte filas repetidas (`duplicated`) y luego otro para contarlas.

Etapa 2.9. Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [16]:
# Elimina los duplicados explícitos
df_cleaned = df.drop_duplicates()

<details>
<summary>Haz clic para ver la pista</summary>

Busca un método que elimine las filas duplicadas directamente del DataFrame. No olvides actualizar la variable.

Etapa 2.10. Comprobemos ahora si conseguimos eliminar todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [17]:
# Comprueba de nuevo si hay duplicados
df_cleaned.duplicated().sum()

0

<div class="alert alert-block alert-success">
<b>Comentario de Revisor </b> <a class="tocSkip"></a>

Perfecto, buen trabajo usando el método <code>drop_duplicates()</code> para solucionar el problema de los valores duplicados en el dataset.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Repite el mismo enfoque que usaste para contar duplicados antes. Deberías ver que el total ahora es cero.

Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Etapa 2.11. Primero debemos mostrar una lista de nombres de géneros únicos, por orden alfabético. Para ello:
1. Extrae la columna `genre` del DataFrame.
2. Llama al método que devolverá todos los valores únicos en la columna extraída.


In [18]:
# Inspecciona los nombres de géneros únicos
sorted_unique_genre = sorted(df_cleaned['genre'].unique())
print(sorted_unique_genre)

['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

<div class="alert alert-block alert-success">
<b>Comentario de Revisor </b> <a class="tocSkip"></a>

Excelente, buen trabajo mostrando los valores únicos de la columna de manera ordenada.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa `unique()` para obtener los géneros únicos. Si quieres verlos en orden alfabético, combina con `sorted()`.

Etapa 2.12. Vamos a examinar la lista para identificar **duplicados implícitos** del género `hiphop`, es decir, nombres mal escritos o variantes que hacen referencia al mismo género musical.

Los duplicados que encontrarás son:

* `hip`  
* `hop`  
* `hip-hop`  

Para solucionarlo, vamos a crear una función llamada `replace_wrong_values()`.


1. Define una función llamada `replace_wrong_values()` que reciba los siguientes parámetros:

* `df`: el DataFrame a modificar
* `column`: el nombre de la columna a trabajar
* `wrong_values`: una lista con los valores incorrectos
* `correct_value`: el valor correcto para reemplazar

2. Dentro de la función, usa un bucle `for` para iterar sobre cada valor incorrecto y aplicar `.replace()`.


In [19]:
# Función para reemplazar los duplicados implícitos
def replace_wrong_values(df, column, wrong_values, correct_value):
    for wrong_value in wrong_values:
        df[column] = df[column].replace(wrong_value, correct_value)
    return(df)

<details>
<summary>Haz clic para ver la pista</summary>

Crea una función con cuatro parámetros. Dentro del bucle, llama a `.replace()` sobre la columna para cada valor incorrecto, cambiándolo por el correcto.

Etapa 2.13. Ahora, llama a la función pasando:

* `df` como el DataFrame
* `'genre'` como nombre de columna
* `['hip', 'hop', 'hip-hop']` como lista de valores incorrectos
* `'hiphop'` como valor correcto


In [20]:
# Elimina los duplicados implícitos

df_cleaned = replace_wrong_values(df_cleaned, 'genre', ['hip', 'hop', 'hip-hop'],'hiphop')
print(df_cleaned)

        user_id                              track            artist  \
0      FFB692EC                  Kamigata To Boots  The Mass Missile   
1      55204538        Delayed Because of Accident  Andreas Rönnberg   
2        20EC38                  Funiculì funiculà       Mario Lanza   
3      A3DD03C9              Dragons in the Sunset        Fire + Ice   
4      E2DC1FAE                        Soul People        Space Echo   
...         ...                                ...               ...   
65074  729CBB09                            My Name            McLean   
65075  D08D4A55  Maybe One Day (feat. Black Spade)       Blu & Exile   
65076  C5E3A0D5                          Jalopiina           unknown   
65077  321D0506                      Freight Train     Chas McDevitt   
65078  3A64EF84          Tell Me Sweet Little Lies      Monica Lopez   

            genre         city      time        day  
0            rock  Shelbyville  20:28:33  Wednesday  
1            rock  Springfi

/tmp/ipykernel_37/1856832352.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[column] = df[column].replace(wrong_value, correct_value)


<div class="alert alert-block alert-success">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

Excelente, buen trabajo estructurando y aplicando la función para solucionar el problema de los valores duplicados implícitos en la columna con información de género.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Llama a la función que creaste y pasa los cuatro argumentos necesarios. Asegúrate de usar el nombre correcto de la columna y lista los valores que deben unificarse.

Etapa 2.14. Asegúrate de que los nombres duplicados se hayan eliminado. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [21]:
# Comprueba de nuevo los duplicados implícitos
sorted_unique_genre = sorted(df_cleaned['genre'].unique())
print(sorted_unique_genre)

['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

<div class="alert alert-block alert-success">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

Buen trabajo. 
</div>


<details>
<summary>Haz clic para ver la pista</summary>

Usa de nuevo el método que muestra los valores únicos de una columna para verificar que todos los nombres se hayan unificado correctamente.

## Etapa 3. Análisis

### Tarea: Comparar el comportamiento de los usuarios en las dos ciudades <a id='activity'></a>

Queremos analizar si hay diferencias en la cantidad de canciones reproducidas en Springfield y Shelbyville. Para ello, usaremos los datos de dos días de la semana: lunes y viernes.

Compararemos cuántas canciones se escucharon en cada ciudad durante esos días para identificar posibles patrones de comportamiento.

Sigue estos tres pasos para organizar tu análisis:

- Dividir: agrupa los datos por ciudad.

- Aplicar: cuenta cuántas canciones se reproducen en cada grupo.

- Combinar: presenta los resultados de forma que se puedan comparar fácilmente ambas ciudades.

Repite este proceso por separado para cada uno de los dos días.

Etapa 3.1.
Cuenta cuántas canciones se reprodujeron en cada ciudad utilizando la columna `'track'` como referencia.

In [22]:
# Cuenta las canciones reproducidas en cada ciudad
df_cleaned.groupby('city')['track'].count()

city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64

<details>
<summary>Haz clic para ver la pista</summary>

Agrupa los datos por ciudad con el método `groupby()` y usa el método `count()` para contar las canciones reproducidas en cada grupo.

Etapa 3.2. Redacta brevemente tus observaciones sobre los resultados.

¿Qué diferencias encontraste entre Springfield y Shelbyville? ¿A qué podrían deberse?


Escribe tus observaciones aquí.
Respuesta: La cantidad de canciones escuchadas en Springfield es mucho mayor a la de las canciones escuchadas en Shelbyville, esto puede deberse a que la cantidad de canciones reproducidas por ciudad es proporcional al número de usuarios por ciudad, la cantidad de usuarios de Shelbyville es mucho menor a la de la cantidad de usuarios de Springfield.

Etapa 3.3.
Agrupa los datos por día de la semana y cuenta cuántas canciones se reprodujeron los lunes y viernes.



In [23]:
# Calcula las canciones reproducidas en cada uno de los dos días

filtered_by_day = df_cleaned.loc[(df_cleaned['day']== 'Monday') | (df_cleaned['day']== 'Friday')]
filtered_by_day.groupby('day')['track'].count()


day
Friday    21840
Monday    21354
Name: track, dtype: int64

<div class="alert alert-block alert-success">
<b>Comentario de Revisor</b> <a class="tocSkip"></a>

Perfecto, buen trabajo estructurando y analizando las dos tablas agrupadas.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Cambia el criterio de agrupación a la columna que representa el día de la semana y filtra solo para los días que te interesan.

Etapa 3.4. Describe brevemente qué observaste al comparar los lunes y viernes.

¿Hubo un día con más actividad? ¿Cambia algo si analizas cada ciudad por separado?

Escribe tus observaciones aquí.

Respuesta:
Se registra más actividad el viernes.
Si, se observa más actividad el viernes que el lunes. Si se analizará por ciudad, si cambiaría el total de canciones reproducidas porque el resultado anterior fue el total de todos los traks escuchados ciertos días de manera general.

Etapa 3.5

Ahora vamos a combinar dos criterios: día y ciudad.

Crea una función llamada `number_tracks()` que reciba dos parámetros:

* `day`: un día de la semana (por ejemplo, `'Monday'`)
* `city`: el nombre de una ciudad (por ejemplo, `'Springfield'`)

Dentro de la función:

1. Filtra el DataFrame por el día.
2. Luego, filtra por la ciudad.
3. Cuenta cuántas veces aparece `'user_id'` en ese filtro.
4. Devuelve ese número como resultado.


In [33]:
# Declara la función number_tracks() con dos parámetros: day= y city=.

def number_tracks(day, city):
    # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=
    day_filter = df_cleaned[df_cleaned['day'] == day] 
    # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=
    city_filter = day_filter[day_filter['city'] == city]
    # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()
    user_id_filter = city_filter['user_id'].count()
    # Devuelve el número de valores de la columna 'user_id'
    return(user_id_filter)
    
  

<details>
<summary>Haz clic para ver la pista</summary>

Define una función con dos parámetros. Dentro, aplica un filtrado secuencial usando `df[...]`, y al final cuenta los valores con `.count()`.

Etapa 3.6. Llama a `number_tracks()` cuatro veces: una por ciudad en cada uno de los dos días.

In [37]:
# El número de canciones reproducidas en Springfield el lunes
monday_result_spr = number_tracks('Monday', 'Springfield')
print(monday_result_spr)

15740


In [35]:
# El número de canciones reproducidas en Shelbyville el lunes
monday_result_shel = number_tracks('Monday', 'Shelbyville')
print(monday_result_shel)

5614


In [38]:
# El número de canciones reproducidas en Springfield el viernes
friday_result_spr = number_tracks('Friday', 'Springfield')
print(friday_result_spr)

15945


In [39]:
# El número de canciones reproducidas en Shelbyville el viernes
friday_result_shel = number_tracks('Friday', 'Shelbyville')
print(friday_result_shel)

5895


<div class="alert alert-block alert-success">
<b>Comentario de Revisor </b> <a class="tocSkip"></a>

Perfecto, buen trabajo estructurando y aplicando la función.

</div>


<details>
<summary>Haz clic para ver la pista</summary>

Llama a tu función con combinaciones distintas de ciudad y día, como `'Monday', 'Springfield'`, y guarda o imprime los resultados para comparar.

# Conclusiones <a id='end'></a>

## Escribe tus conclusiones finales sobre el análisis

Redacta un resumen breve y claro de los hallazgos obtenidos durante el proceso de análisis. Tu conclusión debe:

* Mencionar los principales patrones que observaste en los datos.
* Identificar cualquier problema encontrado y cómo lo solucionaste.
* Explicar cómo estas acciones ayudaron a mejorar la calidad del análisis.

Reflexiona también sobre la pregunta central:

> *¿Los datos muestran que el comportamiento de los usuarios —en cuanto a la música que escuchan— varía según la ciudad y el día de la semana?*

Apoya tu respuesta con ejemplos concretos de los resultados obtenidos.


Detalla aquí tus conclusiones.

Hallazgos: 

La cantidad de usuarios de Springfield es mucho mayor a la de Shelbyville, es por eso que la actividad es notablemente mayor en una que en la otra. Al filtrarlos por días podemos notar que la actividad en ambas ciudades es mayor los viernes en comparación de los lunes en ambas ciudades.

Los datos de primera mano presentaban datos ausentes, variaciones en el uso de mayúsculas y minúsculas, espacios al inicio y datos duplicados; problemas que fueron solucionados con los métodos correspondientes. .lower() para hacer los strings en minúsculas, .strip() para quitar los espacios al inicio, .fillna() para rellenar los datos ausentes con el str 'unknown', .drop_duplicates() para eliminar los datos duplicados. Tambien presentaba variaciones en los nombres de los géneros, los cuales fueron corregidos y reemplazados con el método.replace.

Tomar estas acciones me permitió tener una base más confiable y certera para trabajar. Al estandarizar podemos sacar conclusiones precisas.


¿Los datos muestran que el comportamiento de los usuarios —en cuanto a la música que escuchan— varía según la ciudad y el día de la semana?

Si, los datos muestran que el comportamiento de los usuarios varia según la ciudad y el dia de la semana, por ejemplo:
En total las canciones escuchadas por ciudad es mayor en la ciudad de Springfield y en Shelbyville es menor.
city
Shelbyville    18512
Springfield    42741

Tambien, los resultados varian según el día y la ciudad, por ejemplo:

Los lunes en Springfield se escucharon un total de 15740 canciones
mientras que 
Los lunes en Shelbyville se escucharon un total de 5614 canciones

Los viernes en Springfield se escucharon un total de 15945 canciones
y
Los viernes en Shelbyville se escucharon un total de 5895 canciones

Se puede concluir que en general la cantidad de escuchas es mayor en Springfield que en Shelbyville; pero en ambas ciudades la actividad es mayor los viernes.




<div class="alert alert-block alert-success">
<b>Comentario de Revisor </b> <a class="tocSkip"></a>

Buen trabajo.

</div>
